Что выведет программа? Объяснить вывод программы. Объяснить как работают defer’ы и их порядок вызовов.

```go
package main

import (
	"fmt"
)


func test() (x int) {
	defer func() {
		x++
	}()
	x = 1
	return
}


func anotherTest() int {
	var x int
	defer func() {
		x++
	}()
	x = 1
	return x
}


func main() {
	fmt.Println(test())
	fmt.Println(anotherTest())
}
```

Ответ:
```
2
1

```
`defer` ключевое слово для отложенного выполнения функции, которая выполнится до возвращения управления вызывающей функции. Если в функции объявлено несколько `defer`, то они будут накапливаться в стеке и выполнятся после завершения функции в обратном порядке (LIFO). Если в функции возникнет `panic`, то `defer` всё равно выполнится. Поэтому конструкция `defer` часто используется для перехвата паники и восстановления.

`return` по-умолчанию копирует выходные данные в определённую для этого область памяти, то есть условный `return x` вернёт нам не ту же переменную, а её копию. В случае с именованными возвращаемыми переменными, они уже находятся в памяти для возврата, то есть функция изначально оперирует с данной областью памяти. И если в такой функции есть defer то он так же может изменить эту область памяти. Поэтому `defer` в функции `test()` изменяет именованную возвращаемую переменную `x`. В случае же функции `anotherTest()` `defer` изменит внутреннюю переменную `x`, которая на тот момент уже будет скопирована в выходную область памяти.  

Это, с одной стороны, нежелательное поведение `defer` и его стоит остерегаться. Но этому так же можно найти применение, к примеру это свойство используется при перехвате и восстановлении паники, записывая причину паники в именованную возвращаемую переменную `err`.